File: cogs/warns.py
"""
Warn System Cog
Persistent warns via SQLite. Auto-escalates:
  WARN_MUTE_AT  → timeout 1 hour
  WARN_KICK_AT  → kick
  WARN_BAN_AT   → ban
"""
import discord
from discord.ext import commands
import logging
from config.config import WARN_MUTE_AT, WARN_KICK_AT, WARN_BAN_AT
from utils.utils import is_mod, log_action, mod_embed, utcnow
from db.db import init_db

log = logging.getLogger("modbot.warns")

class Warns(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.Cog.listener()
    async def on_ready(self):
        await init_db()

    @commands.command(name="warn")
    async def warn(self, ctx, member: discord.Member, *, reason: str = "No reason provided"):
        if not is_mod(ctx.author):
            return await ctx.send("❌ No permission.", delete_after=5)
        if member.guild_permissions.administrator:
            return await ctx.send("❌ Cannot warn an admin.", delete_after=5)

        ts = utcnow().isoformat()
        await db.add_warn(ctx.guild.id, member.id, ctx.author.id, reason, ts)
        warns = await db.get_warns(ctx.guild.id, member.id)
        count = len(warns)

        await ctx.send(f"⚠️ {member.mention} warned. Total warns: **{count}**")
        embed = mod_embed(
            "⚠️ Member Warned",
            discord.Color.yellow(),
            user=f"{member} (`{member.id}`)",
            moderator=str(ctx.author),
            reason=reason,
            total_warns=str(count)
        )
        await log_action(ctx.guild, embed)

        # Escalation
        await self._escalate(ctx, member, count)

    async def _escalate(self, ctx, member: discord.Member, count: int):
        import datetime
        if count >= WARN_BAN_AT:
            try:
                await ctx.guild.ban(member, reason=f"Reached {WARN_BAN_AT} warns", delete_message_seconds=0)
                await ctx.send(f"🔨 {member.mention} banned after {WARN_BAN_AT} warns.")
            except discord.Forbidden:
                pass
        elif count >= WARN_KICK_AT:
            try:
                await member.kick(reason=f"Reached {WARN_KICK_AT} warns")
                await ctx.send(f"👢 {member.mention} kicked after {WARN_KICK_AT} warns.")
            except discord.Forbidden:
                pass
        elif count >= WARN_MUTE_AT:
            try:
                await member.timeout(datetime.timedelta(hours=1), reason=f"Reached {WARN_MUTE_AT} warns")
                await ctx.send(f"🔇 {member.mention} timed out (1hr) after {WARN_MUTE_AT} warns.")
            except discord.Forbidden:
                pass

    @commands.command(name="warns")
    async def list_warns(self, ctx, member: discord.Member):
        if not is_mod(ctx.author):
            return await ctx.send("❌ No permission.", delete_after=5)
        warns = await db.get_warns(ctx.guild.id, member.id)
        if not warns:
            return await ctx.send(f"✅ {member.mention} has no warns.")

        embed = discord.Embed(title=f"Warns for {member}", color=discord.Color.yellow())
        for w in warns:
            wid, mod_id, reason, ts = w
            embed.add_field(name=f"#{wid} — {ts[:10]}", value=f"Mod: <@{mod_id}>\nReason: {reason}", inline=False)
        await ctx.send(embed=embed)

    @commands.command(name="unwarn")
    async def unwarn(self, ctx, warn_id: int):
        if not is_mod(ctx.author):
            return await ctx.send("❌ No permission.", delete_after=5)
        await db.remove_warn(warn_id)
        await ctx.send(f"✅ Warn #{warn_id} removed.")

    @commands.command(name="clearwarns")
    async def clearwarns(self, ctx, member: discord.Member):
        if not is_mod(ctx.author):
            return await ctx.send("❌ No permission.", delete_after=5)
        await db.clear_warns(ctx.guild.id, member.id)
        await ctx.send(f"✅ All warns cleared for {member.mention}.")

async def setup(bot):
    await bot.add_cog(Warns(bot))

File: src/handlers/commands/warns.rs
use serenity::{
    all::*,
    framework::standard::{macros::{command, group}, Args, CommandResult},
};
use crate::{BotData, db, utils::{log_action, mod_embed}};
use chrono::Utc;

#[group]
#[commands(warn, warns, unwarn, clearwarns)]
pub struct WarnCmds;

// ── !warn ─────────────────────────────────────────────────────────────────────

#[command]
#[description = "Warn a member. Auto-escalates at configured thresholds."]
pub async fn warn(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
    // Parse target before any async work
    let target_id = match args.single::<UserId>() {
        Ok(id) => id,
        Err(_) => { msg.reply(&ctx.http, "❌ Provide a member mention or ID.").await?; return Ok(()); }
    };
    let reason = args.rest().trim().to_string();
    let reason = if reason.is_empty() { "No reason provided".to_string() } else { reason };

    // Extract all cache data and drop CacheRef before any .await
    let (is_mod, target_is_admin, target_tag, mod_log_channel, guild_id, author_tag) = {
        let data = ctx.data.read().await;
        let cfg = &data.get::<BotData>().unwrap().config;
        let guild_ref = ctx.cache.guild(msg.guild_id.unwrap_or_default());
        let guild = match guild_ref.as_ref() {
            Some(g) => g,
            None => return Ok(()),
        };

        let is_mod = guild.members.get(&msg.author.id)
            .map(|m| cfg.is_mod(m))
            .unwrap_or(false);

        let (target_is_admin, target_tag) = guild.members.get(&target_id)
            .map(|m| (
                m.permissions.map(|p| p.administrator()).unwrap_or(false),
                m.user.tag(),
            ))
            .unwrap_or((false, target_id.to_string()));

        (is_mod, target_is_admin, target_tag, cfg.mod_log_channel, msg.guild_id.unwrap(), msg.author.tag())
    }; // CacheRef dropped here

    if !is_mod {
        msg.reply(&ctx.http, "❌ No permission.").await?;
        return Ok(());
    }
    if target_is_admin {
        msg.reply(&ctx.http, "❌ Cannot warn an admin.").await?;
        return Ok(());
    }

    let ts = Utc::now().to_rfc3339();

    let (pool, warn_mute_at, warn_kick_at, warn_ban_at) = {
        let data = ctx.data.read().await;
        let bd = data.get::<BotData>().unwrap();
        (bd.db.clone(), bd.config.warn_mute_at, bd.config.warn_kick_at, bd.config.warn_ban_at)
    };

    db::add_warn(
        &pool,
        guild_id.get() as i64,
        target_id.get() as i64,
        msg.author.id.get() as i64,
        &reason,
        &ts,
    ).await?;

    let count = db::count_warns(&pool, guild_id.get() as i64, target_id.get() as i64).await?;

    msg.reply(&ctx.http, format!("⚠️ <@{target_id}> warned. Total warns: **{count}**")).await?;

    let embed = mod_embed("⚠️ Member Warned", Color::GOLD, &[
        ("User",        &format!("{target_tag} (`{target_id}`)") as &str),
        ("Moderator",   &author_tag),
        ("Reason",      &reason),
        ("Total Warns", &count.to_string()),
    ]);
    log_action(&ctx.http, mod_log_channel, embed).await;

    // Escalation
    escalate(ctx, guild_id, target_id, count, warn_mute_at, warn_kick_at, warn_ban_at, msg).await;

    Ok(())
}

async fn escalate(
    ctx:          &Context,
    guild_id:     GuildId,
    user_id:      UserId,
    count:        u32,
    warn_mute_at: u32,
    warn_kick_at: u32,
    warn_ban_at:  u32,
    msg:          &Message,
) {
    if count >= warn_ban_at {
        let _ = guild_id.ban_with_reason(&ctx.http, user_id, 0, &format!("Reached {warn_ban_at} warns")).await;
        let _ = msg.reply(&ctx.http, format!("🔨 <@{user_id}> banned after {warn_ban_at} warns.")).await;
    } else if count >= warn_kick_at {
        let _ = guild_id.kick_with_reason(&ctx.http, user_id, &format!("Reached {warn_kick_at} warns")).await;
        let _ = msg.reply(&ctx.http, format!("👢 <@{user_id}> kicked after {warn_kick_at} warns.")).await;
    } else if count >= warn_mute_at {
        let until = (chrono::Utc::now() + chrono::Duration::hours(1)).to_rfc3339();
        let _ = guild_id
            .edit_member(&ctx.http, user_id, EditMember::new().disable_communication_until(until))
            .await;
        let _ = msg.reply(&ctx.http, format!("🔇 <@{user_id}> timed out (1hr) after {warn_mute_at} warns.")).await;
    }
}

// ── !warns ────────────────────────────────────────────────────────────────────

#[command]
#[description = "List warns for a member."]
pub async fn warns(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
    let target_id = match args.single::<UserId>() {
        Ok(id) => id,
        Err(_) => { msg.reply(&ctx.http, "❌ Provide a member mention or ID.").await?; return Ok(()); }
    };

    let (is_mod, _mod_log_channel, guild_id) = {
        let data = ctx.data.read().await;
        let cfg = &data.get::<BotData>().unwrap().config;
        let is_mod = ctx.cache
            .guild(msg.guild_id.unwrap_or_default())
            .and_then(|g| g.members.get(&msg.author.id).map(|m| cfg.is_mod(m)))
            .unwrap_or(false);
        (is_mod, cfg.mod_log_channel, msg.guild_id.unwrap())
    };

    if !is_mod {
        msg.reply(&ctx.http, "❌ No permission.").await?;
        return Ok(());
    }

    let pool = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().db.clone()
    };

    let rows = db::get_warns(&pool, guild_id.get() as i64, target_id.get() as i64).await?;

    if rows.is_empty() {
        msg.reply(&ctx.http, format!("✅ <@{target_id}> has no warns.")).await?;
        return Ok(());
    }

    let mut embed = CreateEmbed::new()
        .title(format!("Warns for <@{target_id}>"))
        .color(Color::GOLD);

    for w in &rows {
        embed = embed.field(
            format!("#{} — {}", w.id, &w.ts[..10.min(w.ts.len())]),
            format!("Mod: <@{}>\nReason: {}", w.mod_id, w.reason),
            false,
        );
    }
    msg.channel_id.send_message(&ctx.http, CreateMessage::new().embed(embed)).await?;
    Ok(())
}

// ── !unwarn ───────────────────────────────────────────────────────────────────

#[command]
#[description = "Remove a specific warn by ID."]
pub async fn unwarn(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
    let warn_id: i64 = match args.single() {
        Ok(id) => id,
        Err(_) => { msg.reply(&ctx.http, "❌ Provide a warn ID.").await?; return Ok(()); }
    };

    let is_mod = {
        let data = ctx.data.read().await;
        let cfg = &data.get::<BotData>().unwrap().config;
        ctx.cache
            .guild(msg.guild_id.unwrap_or_default())
            .and_then(|g| g.members.get(&msg.author.id).map(|m| cfg.is_mod(m)))
            .unwrap_or(false)
    };

    if !is_mod {
        msg.reply(&ctx.http, "❌ No permission.").await?;
        return Ok(());
    }

    let pool = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().db.clone()
    };

    db::remove_warn(&pool, warn_id).await?;
    msg.reply(&ctx.http, format!("✅ Warn #{warn_id} removed.")).await?;
    Ok(())
}

// ── !clearwarns ───────────────────────────────────────────────────────────────

#[command]
#[description = "Clear all warns for a member."]
pub async fn clearwarns(ctx: &Context, msg: &Message, mut args: Args) -> CommandResult {
    let target_id = match args.single::<UserId>() {
        Ok(id) => id,
        Err(_) => { msg.reply(&ctx.http, "❌ Provide a member mention or ID.").await?; return Ok(()); }
    };

    let (is_mod, guild_id) = {
        let data = ctx.data.read().await;
        let cfg = &data.get::<BotData>().unwrap().config;
        let is_mod = ctx.cache
            .guild(msg.guild_id.unwrap_or_default())
            .and_then(|g| g.members.get(&msg.author.id).map(|m| cfg.is_mod(m)))
            .unwrap_or(false);
        (is_mod, msg.guild_id.unwrap())
    };

    if !is_mod {
        msg.reply(&ctx.http, "❌ No permission.").await?;
        return Ok(());
    }

    let pool = {
        let data = ctx.data.read().await;
        data.get::<BotData>().unwrap().db.clone()
    };

    db::clear_warns(&pool, guild_id.get() as i64, target_id.get() as i64).await?;
    msg.reply(&ctx.http, format!("✅ All warns cleared for <@{target_id}>.")).await?;
    Ok(())
}
